### PR TITLE
Graph dropdown

### PR DESCRIPTION
**Pull request by kenneth-miura**
## Motivation and Context

Implementing graph dropdown, as specified in issue #1218 
## Your Changes
Implemented a graph dropdown component, mimicking the way `InfoBox` is implemented. Removed Graphs from sidebar.

Also updated tests to reflect changes to Sidebar & addition of properties to graph. Added test for new graph dropdown component

**Description**:

**Type of change** (select all that apply):

- [x] User Interface Change

## Testing

Manual. Please note that the GraphDropdown tests are currently failing.
## Questions and Comments (if applicable)

I'm having trouble testing features that are related to hovering over the "Graph" element in the nav-bar. Since it's not rendered in React, I don't know how to get access to it in order to use it in tests. I tried using `TestContainer`, but the `Container` is a sibling of a navbar so it didn't work.

Also please ignore the strange formatting on some of these files. Not sure where it came from, will fix.
## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have verified that the CircleCI tests have
  passed. 
- [x] I have added tests for my changes.

### CODE REVIEW

**Comment by kenneth-miura** [comment]
I can't figure out how to get the tests working with a hard-coded nav bar, so I'm going to work on other stuff. I've removed the tests that require the nav bar from the GraphDropdown Tests, and changed the code so it passes all the pre-existing tests.

**Review by david-yz-liu — COMMENTED** [review]
@kenneth-miura great work! A few overall comments about this PR:

1. It would be nice to add a "down arrow/triangle" icon to the graph link in the nav bar when displaying the graph page. This would make it more obvious that there's more to see when hovering over the Graph link on this page (since the other links don't have this, and this wouldn't appear for the Graph link on other pages).
2. In terms of CSS code style, it is good to have two different classes, but in order to reduce code duplication, it's better to have one class that's the "main" styling for the component, and then a second which is "hidden", which sets a `display: none` rule to hide the component. (Ideally this class would be generic enough to reuse across different components).
3. Also, in your CSS rules please vertically align the right-hand sides (similar to the existing rules).
4. In terms of the UI styling, use left-aligned text, add some additional whitespace between each list item, and remove the left padding (this comes by default on lists). Not a big fan of the background colour, maybe poll the other Courseography folks for alternate suggestions.

**Comment by kenneth-miura** [comment]
For point 2, I used `opacity` over `display:none` so it would do transitions. I will trim the css down to the bare minimum requirements to avoid accidentally selecting things from the dropdown while it's supposed to be hidden.

**Comment by kenneth-miura** [comment]
Before:
image

After:
image

File: js/components/graph/Graph.js [hunk 1]
```diff
@@ -426,6 +435,19 @@ export default class Graph extends React.Component {
...
+    this.setState({showGraphDropdown: true})
+  }
+
+  hideGraphDropdown =  () => {
+    var timeout = setTimeout(() => {
+      this.setState({showGraphDropdown: false})
+    }, 500);
+    this.setState({ timeouts: this.state.timeouts.concat(timeout)})
+  }
+
```
Comment (david-yz-liu) [inline js/components/graph/Graph.js:449]: Delete this blank line

File: js/components/graph/Graph.js [hunk 2]
```diff
@@ -426,6 +435,19 @@ export default class Graph extends React.Component {
...
+  setShowGraphDropdown = () => {
+    this.clearAllTimeouts();
+    this.setState({showGraphDropdown: true})
+  }
+
+  hideGraphDropdown =  () => {
+    var timeout = setTimeout(() => {
+      this.setState({showGraphDropdown: false})
+    }, 500);
+    this.setState({ timeouts: this.state.timeouts.concat(timeout)})
```
Comment (david-yz-liu) [inline js/components/graph/Graph.js:447]: Delete the space after the `{`

File: js/components/graph/Graph.js [hunk 3]
```diff
@@ -426,6 +435,19 @@ export default class Graph extends React.Component {
     });
   };
 
+  setShowGraphDropdown = () => {
+    this.clearAllTimeouts();
+    this.setState({showGraphDropdown: true})
```
Comment (david-yz-liu) [inline js/components/graph/Graph.js:440]: Add a semicolon here (and on the other lines you added below in this file)

File: js/components/graph/GraphDropdown.js [hunk 4]
```diff
@@ -0,0 +1,104 @@
+import React from "react";
+import PropTypes from "prop-types";
+
+export default class GraphDropdown extends React.Component{
+
+	generateDropdownItems(column){
```
Comment (david-yz-liu) [inline js/components/graph/GraphDropdown.js:6]: After reflecting on this a bit, I don't think Courseography should be imposing categories on the graphs. Revert this change to the two-column layout you had before, and do not use any headings. This should greatly simplify this component.

File: js/components/graph/GraphDropdown.js [hunk 5]
```diff
@@ -0,0 +1,104 @@
+import React from "react";
```
Comment (david-yz-liu) [inline js/components/graph/GraphDropdown.js:1]: This entire file seems to be using tabs instead of spaces. Make sure to configure your text editor to use spaces, and then convert the indentation in this file.

File: js/components/graph/GraphDropdown.js [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+				<h4>(Unofficial) Languages</h4>
+				{this.generateDropdownItems(languagesColumn)}
+			</div>
+		</ul>
+
+		);
+	}
+
+	render() {
+		let halfComponentWidth = 412;
```
Comment (david-yz-liu) [inline js/components/graph/GraphDropdown.js:88]: These numbers shouldn't be hard-coded. The `Graph` component computes its width, and you can pass that information as a prop to this component.

For `graphTabLeft`, there again should not be a default number. If the nav bar item can't be found, it's better not to display anything rather than display a component that doesn't make sense.

File: js/components/graph/GraphDropdown.js [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+	}
+}
+
+GraphDropdown.propTypes = {
+	showGraphDropdown: PropTypes.bool,
+	onMouseMove: PropTypes.func,
+	onMouseLeave: PropTypes.func,
+	graphs: PropTypes.array,
+	updateGraph: PropTypes.func
+}
```
Comment (david-yz-liu) [inline js/components/graph/GraphDropdown.js:104]: Make sure to end this file with a blank line.

File: js/components/graph/GraphDropdown.js (hunk 5 above)
Comment (david-yz-liu) [reply js/components/graph/GraphDropdown.js:1]: The same is true to the GraphDropdown test file below.

File: js/components/graph/__tests__/GraphDropdown.test.js [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+		const graphDropdown  = container.getByTestId("test-graph-dropdown");
+		expect(graphDropdown.classList.contains("graph-dropdown-hidden")).toBe(true);
+		fireEvent(graphNav, mouseEnter);
+		expect(graphDropdown.classList.contains("graph-dropdown-display")).toBe(true);
+
+		fireEvent(graphNav, mouseLeave);
+		await sleep(600);
+		expect(graphDropdown.classList.contains("graph-dropdown-hidden")).toBe(true);
+	});
+});
```
Comment (david-yz-liu) [inline js/components/graph/__tests__/GraphDropdown.test.js:76]: Make sure to end this file with a blank line.

File: style/app.scss [hunk 9]
```diff
@@ -2487,14 +2480,44 @@ main-filter input
 
 .tooltip-group-display
 {
-  cursor: pointer;
-  transition: opacity 0.4s linear;
-  opacity:1;
+  cursor     : pointer;
```
Comment (david-yz-liu) [inline style/app.scss:2483]: Revert the style changes for this rule and the one below. You can make these in a separate PR.

File: style/app.scss [hunk 10]
```diff
@@ -2487,14 +2480,44 @@ main-filter input
...
+}
+
+.graph-dropdown-hidden
+{
+  opacity    : 0;
+  position   : absolute;
+  display    : inline-block;
+  height     : 0;
+  overflow   : hidden;
+  transition : opacity 0.4s linear;
```
Comment (david-yz-liu) [inline style/app.scss:2502]: I don't like these transitions. You can keep the timeout in the JS code, but remove both this rule and the one for graph-dropdown-display. Use the display rule.

File: style/app.scss [hunk 11]
```diff
@@ -2487,14 +2480,44 @@ main-filter input
...
+  text-align: left;
+  border-bottom: 10px solid #8a75ae;
+  position:absolute;
+}
+.graph-dropdown-display h4
+{
+  font-weight: bold;
+  margin-top: 5px;
+  margin-bottom: 5px;
+}
```
Comment (david-yz-liu) [inline style/app.scss:2523]: Make sure this file ends in a blank line. In fact, you should configure your text editor to always do this automatically.

File: style/app.scss [hunk 12]
```diff
@@ -2487,14 +2480,44 @@ main-filter input
...
+  opacity          : 1;
+  transition       : opacity 0.4s linear;
+  list-style       : none;
+  padding-left     : 20px;
+  column-count: 3;
+  text-align: left;
+  border-bottom: 10px solid #8a75ae;
+  position:absolute;
+}
+.graph-dropdown-display h4
```
Comment (david-yz-liu) [inline style/app.scss:2518]: Delete this rule, it's no longer necessary after you address my above comments.

File: style/app.scss [hunk 13]
```diff
@@ -2487,14 +2480,44 @@ main-filter input
...
+.graph-dropdown-display
+{
+  background-color : #8a75ae;
+  font-size        : 15px;
+  display          : inline-block;
+  opacity          : 1;
+  transition       : opacity 0.4s linear;
+  list-style       : none;
+  padding-left     : 20px;
+  column-count: 3;
```
Comment (david-yz-liu) [inline style/app.scss:2513]: Fix the style on these rules.

File: style/app.scss [hunk 14]
```diff
@@ -837,21 +837,14 @@ a:active
...
-  width            : 90%;
-  background-color : #ffffff;
-  margin-bottom    : 20px;
+  text-indent: 2em;
 }
 
-.graph-button:hover
+.graph-dropdown-item:hover
 {
+  cursor           : pointer;
```
Comment (david-yz-liu) [inline style/app.scss:847]: This rule should be on the item (see above), not just on `:hover`.

File: style/app.scss [hunk 15]
```diff
@@ -837,21 +837,14 @@ a:active
...
-  display          : block;
-  cursor           : pointer;
-  font-size        : 20px;
-  border           : solid 1px #000000;
-  border-radius    : 6px;
-  text-align       : center;
-  width            : 90%;
-  background-color : #ffffff;
-  margin-bottom    : 20px;
+  text-indent: 2em;
```
Comment (david-yz-liu) [inline style/app.scss:842]: Delete this rule.

**Comment by david-yz-liu** [comment]
> For point 2, I used `opacity` over `display:none` so it would do transitions. I will trim the css down to the bare minimum requirements to avoid accidentally selecting things from the dropdown while it's supposed to be hidden.

@kenneth-miura As I mentioned in my second review, I don't like the transition there. So you should do as I asked in Point 2 in my original review, in addition to fixing the things I mentioned in my second review.

File: js/components/graph/GraphDropdown.js (hunk 6 above)
Comment (kenneth-miura) [reply js/components/graph/GraphDropdown.js:88]: The halfComponentWidth is based off half the width of the GraphDropdown component. I couldn't figure out how to get GraphDropdown centered on the graph tab in css. 

I think with the text aligned to the left it looks good without trying to center GraphDropdown on the graph tab (screenshot below). Are you ok with that? If not, I think there's the translation property in CSS which I could use to center the graph dropdown, so I can just do that.

image

Comment (david-yz-liu) [reply js/components/graph/GraphDropdown.js]: @kenneth-miura oh I see, yes I think aligning it to the left looks good, you can stick with that!

(BTW, make sure to add some padding to the dropdown container, so that the text isn't right against the borders).

**Comment by kenneth-miura** [comment]
Sorry, forgot to include screenshot with changes. Here is what I have right now, post adding the padding
image
